File: 0064-minimum-path-sum/0064-minimum-path-sum.java
class Solution {
    public int minPathSum(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        int[][] dp = new int[m + 1][n + 1];        
        for (int i = 0; i <= m; i++) {
            Arrays.fill(dp[i], Integer.MAX_VALUE);
        }
        dp[m - 1][n - 1] = grid[m - 1][n - 1];
        for (int i = m - 1; i >= 0; i--) {
            for (int j = n - 1; j >= 0; j--) {
                if (i == m - 1 && j == n - 1) continue;
                dp[i][j] = Math.min(dp[i+1][j], dp[i][j+1]) + grid[i][j];
            }
        }
        return dp[0][0];
        /**
            [[1,3,1],
             [1,5,1],
             [4,2,1]] 
             
             dp[i][j] = min(r,d) + grid[i][j]
                * ->
                V
              1 3
              1 5
                        ij
                    i+1,j   i,j+1

                            00
                        /         \
                      10             01
                   /    5\        /     \ 
                 ~      11      11       ~
                 dp[i][j]
                 return min(d, r) + curr
         */
    }
}

File: 0064-minimum-path-sum/0064-minimum-path-sum.py
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        M = len(grid)
        N = len(grid[0])
        dp = [0] * N        
        for r in range(M - 1, -1, -1):
            for c in range(N - 1, -1, -1):
                if (c + 1) == N:
                    dp[c] = grid[r][c] + dp[c]
                elif (r + 1) == M:
                    dp[c] = grid[r][c] + dp[c + 1]
                else:
                    dp[c] = grid[r][c] + min(dp[c], dp[c + 1])

        return dp[0]